includes/class-wc-payments-express-checkout-button-display-handler.php

```diff
@@ -0,0 +1,102 @@
+ * @package WooCommerce\Payments
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Class WC_Payments_Express_Checkout_Button_Display_Handler
+ */
+class WC_Payments_Express_Checkout_Button_Display_Handler {
+
+	/**
+	 * WC_Payment_Gateway_WCPay instance.
+	 *
+	 * @var WC_Payment_Gateway_WCPay
+	 */
+	private $gateway;
+
+	/**
+	 * Instance of WC_Payments_Payment_Request_Button_Handler, created in init function
+	 *
+	 * @var WC_Payments_Payment_Request_Button_Handler
+	 */
+	private $payment_request_button_handler;
+
+	/**
+	 * Instance of WC_Payments_Platform_Checkout_Button_Handler, created in init function
+	 *
+	 * @var WC_Payments_Platform_Checkout_Button_Handler
+	 */
+	private $platform_checkout_button_handler;
+
+	/**
+	 * Initialize class actions.
+	 *
+	 * @param WC_Payment_Gateway_WCPay                     $gateway WCPay gateway.
+	 * @param WC_Payments_Payment_Request_Button_Handler   $payment_request_button_handler Payment request button handler.
+	 * @param WC_Payments_Platform_Checkout_Button_Handler $platform_checkout_button_handler Platform checkout button handler.
+	 */
+	public function __construct( WC_Payment_Gateway_WCPay $gateway, WC_Payments_Payment_Request_Button_Handler $payment_request_button_handler, WC_Payments_Platform_Checkout_Button_Handler $platform_checkout_button_handler ) {
+		$this->gateway                          = $gateway;
+		$this->payment_request_button_handler   = $payment_request_button_handler;
+		$this->platform_checkout_button_handler = $platform_checkout_button_handler;
+
+		$this->platform_checkout_button_handler->init();
+		$this->payment_request_button_handler->init();
+
+		$is_woopay_express_checkout_enabled = WC_Payments_Features::is_woopay_express_checkout_enabled();
+
+		if ( $is_woopay_express_checkout_enabled || $this->gateway->get_option( 'payment_request' ) ) {
+			add_action( 'woocommerce_after_add_to_cart_quantity', [ $this, 'display_express_checkout_buttons' ], 1 );
+			add_action( 'woocommerce_proceed_to_checkout', [ $this, 'display_express_checkout_buttons' ], 1 );
+			add_action( 'woocommerce_checkout_before_customer_details', [ $this, 'display_express_checkout_buttons' ], 1 );
+
+			if ( $this->gateway->get_option( 'payment_request' ) ) {
+				// Load separator on the Pay for Order page.
+				add_action( 'before_woocommerce_pay_form', [ $this, 'display_express_checkout_buttons' ], 1 );
+			}
+		}
+	}
+
+	/**
+	 * Display express checkout separator only when express buttons are displayed.
+	 *
+	 * @return void
+	 */
+	public function display_express_checkout_separator_if_necessary() {
+		$woopay          = $this->platform_checkout_button_handler->is_woopay_enabled() && $this->platform_checkout_button_handler->should_show_platform_checkout_button();
+		$payment_request = $this->payment_request_button_handler->should_show_payment_request_button();
+		$should_hide     = $payment_request && ! $woopay;
+		if ( $woopay || $payment_request ) {
```

Comment:
```suggestion
		if ( $should_show_woopay || $should_show_payment_request ) {
```
This is a really minor quibble and one you're welcome to ignore, but I'd probably use more descriptive variable names here. 🤷

Reply:
```suggestion
		if ( ! $should_hide ) {
```
Actually, shouldn't we just check for this instead? Is there any reason to add this HTML to the DOM, if we are going to be hiding it?

Reply:
>$should_show_woopay || $should_show_payment_request

I'm fine with these name changes. Thanks for the suggestion!

>Actually, shouldn't we just check for this instead? Is there any reason to add this HTML to the DOM, if we are going to be hiding it?

The `$should_hide` flag is only true when Payment Request is enabled and WooPay express checkout is not. We do indeed have to add this HTML in this case, but we hide it by default and show it only when the browser has a Google Pay or Apple Pay payment method stored. There's more on this [here](https://github.com/Automattic/woocommerce-payments/pull/5399#discussion_r1073633776).